### PR TITLE
Feature x-validate

### PR DESCRIPTION
**Pull request by SimoTod**

### CODE REVIEW

**Comment by KevinBatdorf** [comment]
You need any parts of this worked on? Going to visit my parents in a few hours but I should have time tomorrow to work on anything.

**Comment by SimoTod** [comment]
Checkbox groups, any other rules you can think it's a common case, a hook to define custom rules and tests, but I'm not in a rush

**Comment by KevinBatdorf** [comment]
Here's some others I added to another project last year. Do we want any of these?

```js
const validations = {
  size: (v, t) => String(v).length == t,
  between: (v, t) => { // size:4|between:1920,2002
    let edge = t.split(',')
    return edge[0] <= v && v <= edge[1]
  },
  in: (v, t) => t.split(',').includes(v), // 'in:amb,blu,brn,gry,grn,hzl,oth',
}
```
Reference: https://github.com/KevinBatdorf/advent-2020-javascript-challenges/blob/master/day4/day4.js

**Comment by SimoTod** [comment]
Yeah, I think they have valid use cases. We have minlength and maxlength, so maybe we can call it `length` instead of size (easier to remember).

**Comment by KevinBatdorf** [comment]
I can set aside some time this week if you want me to work on something here. Or even just take a close look at it.

**Comment by SimoTod** [comment]
> I can set aside some time this week if you want me to work on something here. Or even just take a close look at it.

@KevinBatdorf I should be at a good point now. I still have a few tests to implement but the basic support is there (including checkbox arrays and select multiple). I need to write the documentation and proper examples as well but if you want to start having a look or improve something, add more rules/tests, etc, you are more than welcome.

**Comment by KevinBatdorf** [comment]
I went through some of this on the stream (attempt). Some questions I have:

1. `match` might be better named as `equals`? And we could also have `looslyEquals`?
2. `x-validate.immediate` maybe didn't do what I expected. I thought it would trigger other `$invalid()`s to render. I didn't dive too deep into it though.
3. `$invalid($refs.req, 'required')` I thought would accept an array as the second param, but maybe that's not super useful anyway.

Someone stopped by on my stream, so I got cut off early. I'll try to go through it a bit tomorrow as well. I also thought Twitch would have saved the video but I guess that's up to me. Ryan said the video froze up though and I think he gave up on me so I don't know if I'm meant to be in the streaming game anyway :)

**Comment by SimoTod** [comment]
> 1. `match` might be better named as `equals`? And we could also have `looslyEquals`?
Equals sounds good. I don't know if we need loosely equals, values are always strings.

> 2. `x-validate.immediate` maybe didn't do what I expected. I thought it would trigger other `$invalid()`s to render. I didn't dive too deep into it though.
It's in the readme (probably it can be worded in a better way). Basically it triggers the validation on input rather then on blur. 

> 3. `$invalid($refs.req, 'required')` I thought would accept an array as the second param, but maybe that's not super useful anyway.
At the moment, only strings. The idea is that you can write different messages for different rules. 

> Someone stopped by on my stream, so I got cut off early. I'll try to go through it a bit tomorrow as well. I also thought Twitch would have saved the video but I guess that's up to me. Ryan said the video froze up though and I think he gave up on me so I don't know if I'm meant to be in the streaming game anyway :) 
Haha, I think it's just a matter of practice. Next time will be better.

**Comment by KevinBatdorf** [comment]
Sounds good. Do you want more validators written for anything? I can work on something over the weekend. Will try to stream it again but hard to find the time

**Comment by SimoTod** [comment]
Ready for review

File: src/validate.js [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+                    : (
+                        // Note Alpine has a bug end it strips all backslashes breaking regular expresions
+                        (typeof target.$valid === 'string')
+                        ? target.$valid.replace(/\\/g, '') === errorType
+                        : target.$valid === errorType
+                    )
+            }
+        })
+
+        Alpine.onBeforeComponentInitialized((component) => {
```
Comment (ryangjchandler) [inline src/validate.js:74]: Could we maybe wrap all of this up into a small helper function? The part that handles whether or not the element has the specified attribute, etc.

**Review by ryangjchandler — CHANGES_REQUESTED** [review]

File: src/validate.js (hunk 1 above)
Comment (SimoTod) [reply src/validate.js:74]: Sure, we can send another PR to master and then update this one to keep the feature separated. I won't have time to work on it this week though but if any of you guys have time to create the helper, feel free to give it a go.

**Review by KevinBatdorf — APPROVED** [review]

**Pull request closed without merging**